web/api/v1/codec.go

```diff
@@ -13,18 +13,22 @@
 
 package v1
 
-import "github.com/munnerz/goautoneg"
+import (
+	"net/http"
+
+	"github.com/munnerz/goautoneg"
+)
 
 // A Codec performs encoding of API responses.
 type Codec interface {
 	// ContentType returns the MIME time that this Codec emits.
 	ContentType() MIMEType
 
 	// CanEncode determines if this Codec can encode resp.
-	CanEncode(resp *Response) bool
+	CanEncode(req *http.Request, resp *Response) bool
 
 	// Encode encodes resp, ready for transmission to an API consumer.
-	Encode(resp *Response) ([]byte, error)
+	Encode(req *http.Request, resp *Response) ([]byte, error)
```

Comment:
I expect to change either `CanEncode` or `Encode` but not both of them.

Reply:
Changing one of them is enough for this use case but having req in `Encode` could be useful in the future

Reply:
I'd agree with @yeya24 and only add it to `CanEncode` to help with the Thanos use case, to avoid making the code diverge for handling particular special cases.

Reply:
Can we still keep the request in `Encode`? I do think we have a usecase to customize the response based on request in `Encode`. Although it is doable to always create a new Codec but having the request here seems easider to me.

Reply:
@yeya24 Is there a specific usecase you are thinking of?

Reply:
I don't agree with making an unnecessary change.